### PR TITLE
Iterating through Leaves

### LINKED ISSUE #3: Iterate Trie Leaves
Implement an analogue to keys_with_prefix;
https://github.com/casper-network/casper-node/blob/7881d1d6638a8c4fab447b47dfb8b957ef77f8bd/execution_engine/src/storage/global_state/mod.rs#L50-L55

### PR DESCRIPTION
**Pull request by xcthulhu**
Closes #3

### CODE REVIEW

File: src/store/updater.rs [hunk 1]
```diff
@@ -208,39 +214,39 @@ fn empty_branches() -> [UpdatingTrie; 256] {
...
         let data = &data[affix_length + 1..];
 
         let mut branches = empty_branches();
-        let branch_count = match tag_code {
-            NODE31_TAG => {
+        let branch_count = match tag {
+            Tag::Node31 => {
                 // Branch count is the lower 5 bits of the first byte (the higher bits are the tag)
                 let branch_count = (tag_code_and_branch_count & 0b11111) as usize;
                 for idx in 0..branch_count {
```
Comment (afck) [inline src/store/updater.rs:252]: Maybe simpler:

```suggestion
                for branch in data[branch_count..].chunks(DIGEST_LENGTH).take(branch_count) {
```

File: src/store.rs [hunk 2]
```diff
@@ -97,59 +65,38 @@ impl<'a> Iterator for TrieLeavesUnderPrefixIterator<'a> {
...
-            self.prefix.clear();
+            self.initialized = true;
             self.node_stack.push((current_lookup_digest, 0))
         }
 
         while let Some((node_digest, branch_idx)) = self.node_stack.pop() {
-            let trie = match self.store.get_trie_ref(&node_digest) {
+            let trie = match self.store.get_trie(&node_digest) {
                 Some(trie) => trie,
                 None => {
```
Comment (afck) [inline src/store.rs:86]: That shouldn't happen in a consistent trie, should it?

File: src/store.rs [hunk 3]
```diff
@@ -97,59 +65,38 @@ impl<'a> Iterator for TrieLeavesUnderPrefixIterator<'a> {
...
+                Ok(TrieLeafOrBranch::Leaf(leaf)) => {
                     return Some(leaf);
                 }
-                Ok(TrieRead::DigestReadFromNode(new_digest)) => {
+                Ok(TrieLeafOrBranch::Branch(new_digest)) => {
                     self.node_stack.push((node_digest, branch_idx + 1));
-                    self.node_stack.push((new_digest.clone(), 0));
+                    self.node_stack.push((*new_digest, 0));
                 }
+                Ok(TrieLeafOrBranch::IndexOutOfRange | TrieLeafOrBranch::KeyNotFound) => continue,
```
Comment (afck) [inline src/store.rs:99]: `IndexOutOfRange` is normal, but `KeyNotFound` also shouldn't happen, right?

**Review by afck — COMMENTED** [review]
Looks great so far! :+1:

File: src/store.rs [hunk 4]
```diff
@@ -1,94 +1,62 @@
...
-        };
-        TrieLeavesUnderPrefixIterator {
-            root,
-            prefix,
-            store: &self,
-            node_stack,
-        }
-    }
+pub trait TrieStore {
+    fn get_trie(&self, digest: &Digest) -> Option<Trie>;
```
Comment (dwerner) [inline src/store.rs:19]: I think this should have an associated type `Error` and then return `Result<_, Self::Error>` instead. Logic errors within the trie as well as underlying data store errors need to be supported.

File: src/store.rs (hunk 3 above)
Comment (dwerner) [reply src/store.rs:99]: KeyNotFound should be a `Self::Error`, in alignment with my suggestion above?

File: src/store/updater.rs [hunk 5]
```diff
@@ -44,10 +50,10 @@ pub struct KeyMustHaveAtMost255Bytes {
 impl TrieLeafRef {
     pub(crate) fn new(key: Vec<u8>, value: Vec<u8>) -> Result<Self, KeyMustHaveAtMost255Bytes> {
         let key_byte_count = key.len();
-        if key_byte_count > 255 {
+        if key_byte_count > MAX_KEY_BYTES_LEN as usize {
```
Comment (dwerner) [inline src/store/updater.rs:53]: :+1:

File: src/lib.rs [hunk 6]
```diff
@@ -252,5 +276,63 @@ mod tests {
 
             assert_eq!(insert_all_at_once_digest, insert_and_insert_again_digest)
         }
+
+        #[proptest]
+        fn prefix_iterator(prefix: Vec<u8>, keys: Vec<[u8; 10]>) {
+            let mut keys = keys;
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/lib.rs:282]: Looks like the `proptest` macro isn't clever enough to support this... :-(

File: src/lib.rs [hunk 7]
```diff
@@ -252,5 +276,63 @@ mod tests {
...
+            let expected: Vec<_> = keys
+                .into_iter()
+                .filter(|key| key.starts_with(&prefix))
+                .collect();
+            let actual: Vec<[u8; 10]> = store
+                .leaves_under_prefix(root, &prefix)
+                .map(|leaf| {
+                    leaf.key()
+                        .try_into()
+                        .expect("Could not convert trie leaf key to [u8; 5]")
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/lib.rs]: Fixed in 9b453250eb3dd70bd5458d42d40f88fd843ad4af

File: src/lib.rs [hunk 8]
```diff
@@ -252,5 +276,63 @@ mod tests {
...
+                let mut updater = Updater::new(&mut store, EMPTY_DIGEST);
+                for key in &keys {
+                    updater.put(key.to_vec(), vec![0u8]).expect("Could not put");
+                }
+                updater.commit()
+            };
+
+            let expected: Vec<_> = keys
+                .into_iter()
+                .filter(|key| key.starts_with(&prefix))
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/lib.rs:297]: Since this is a `proptest`; it's not exactly random; once a failing test is found proptest will try to *shrink* the failing input into something minimal.  You can read more about this in [shrinking basics][1] in the proptest book if you are interested.

[1]: https://altsysrq.github.io/proptest-book/proptest/tutorial/shrinking-basics.html

File: src/store/updater.rs [hunk 9]
```diff
@@ -44,10 +50,10 @@ pub struct KeyMustHaveAtMost255Bytes {
 impl TrieLeafRef {
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/store/updater.rs]: Fixed in dc79b080a66c3b41e37b2a9bfd8c83bcff7e4fad

File: src/store/updater.rs [hunk 10]
```diff
@@ -208,39 +214,39 @@ fn empty_branches() -> [UpdatingTrie; 256] {
 
 #[derive(thiserror::Error, Debug)]
 pub enum TrieToFancyTrieConversionError {
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/store/updater.rs]: Fixed in dc79b080a66c3b41e37b2a9bfd8c83bcff7e4fad

File: src/wire_trie.rs [hunk 11]
```diff
@@ -189,18 +194,18 @@ impl<'a> WireTrieRef<'a> {
             return None;
         }
 
-        let mut digest_index: usize = 0;
-        // TODO: unravel these loops / avx popcount
+        let mut digest_index: u8 = 0;
+        // TODO: unravel these loops / avx popcount?
         // https://github.com/WojciechMula/sse-popcount/blob/master/popcnt-avx512-vpopcnt.cpp https://arxiv.org/abs/1611.07612
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/wire_trie.rs]: Comment removed in 87e7a7c597d44f1a1db5d9f5ac396fc41437596e

File: src/wire_trie.rs [hunk 12]
```diff
@@ -189,18 +194,18 @@ impl<'a> WireTrieRef<'a> {
...
             // Convert blocks of 32 bytes into u64s, call popcnt to get how many branches in this
             // bit-array part
             let branch_count_in_bit_array =
                 u64::from_le_bytes(branch_byte_indices[i * 8..i * 8 + 8].try_into().ok()?);
-            digest_index += branch_count_in_bit_array.count_ones() as usize;
+            digest_index += branch_count_in_bit_array.count_ones() as u8;
         }
 
-        digest_index += (highest_bits & !(!0u64 << rem.saturating_sub(1))).count_ones() as usize;
+        digest_index += (highest_bits & !(!0u64 << rem.saturating_sub(1))).count_ones() as u8;
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/wire_trie.rs]: I went ahead and made this change in fb56628d8f47ca275077b2b8d14879550f76c5e2 .

I'm still scratching my head here: code coverage says this is 100% covered and the tests are passing either way.  This merits more investigation.

File: src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (afck) [reply src/lib.rs:337]: Yes, but is it likely it will find an input where it _is_ a prefix of one of the keys in the first place? How many keys does it pick by default, and how long a prefix?

E.g. if it always starts with a ~10-byte prefix and 100 keys, it will almost never try a key that extends the prefix. So the test will almost never fail (even if the code were broken) and it won't even start shrinking the inputs.

File: src/wire_trie.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (afck) [reply src/wire_trie.rs]: It would only fail if two neighboring bits in the same `u64` were both `1` and had different values. I think if bits number 2 and 3 are `1` and bit number 1 is `0` then `find_branch_byte(3)` would have returned digest 2. `find_branch_byte(2)` would _also_ have output digest 2, and `find_branch_byte(1)` would correctly have said `None`.

**Review by afck — APPROVED** [review]
My blockers have been addressed.
But I guess it would be good to have more tests for `find_branch_byte`.

File: src/wire_trie.rs [hunk 13]
```diff
@@ -212,16 +217,19 @@ impl<'a> WireTrieRef<'a> {
...
     }
 
-    pub(crate) fn get_nth_digest(&self, n: u8) -> Result<TrieRead<'a>, TrieReadError> {
-        if self.tag() == LEAF_TAG {
-            return Ok(TrieRead::TrieLeaf(WireTrieLeafRef(&self.0[1..])));
+    pub(crate) fn get_nth_digest(
+        &self,
+        digest_index: u8,
+    ) -> Result<TrieLeafOrBranch<'a>, TrieReadError> {
+        if self.tag() == Tag::Leaf {
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/wire_trie.rs:222]: In this conversations I think we should go with your design to have values associated with nodes and get rid of leaves altogether.

File: src/lib.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/lib.rs:336]: It does!  It starts with small vectors including `vec![]`, which caught corner cases I later made into unit tests.  It will also find simple failing test examples where keys are mostly `[0,0,...]`.

But I made another prop test with `[bool ; 10]` keys in 52e9d2747c98cee7917e250e219ce114c814d814

File: src/store.rs [hunk 14]
```diff
@@ -1,94 +1,62 @@
...
+        if !self.initialized {
+            let mut current_lookup_digest = self.root;
             let mut prefix_bytes_read: u8 = 0;
             loop {
-                let trie = match self.store.get_trie_ref(&current_lookup_digest) {
+                let trie = match self.store.get_trie(&current_lookup_digest) {
                     Some(trie) => trie,
                     None => {
-                        // Could not look up digest. The trie-store is likely corrupted.
+                        // Could not look up digest. The trie-store may be corrupted.
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/store.rs]: I've tried to clear this up in 892340e78339fbf754285b7487376ce1681a90ed

**Review by dwerner — APPROVED** [review]

File: src/store/backends/in_memory.rs [hunk 15]
```diff
@@ -0,0 +1,48 @@
...
+        root: Digest,
+        prefix: &'b [u8],
+    ) -> TrieLeavesUnderPrefixIterator<'a, 'b, Self> {
+        TrieLeavesUnderPrefixIterator {
+            root,
+            store: self,
+            prefix,
+            initialized: false,
+            node_stack: vec![],
+        }
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/store/backends/in_memory.rs]: I introduced a `new` constructor for `TrieLeavesUdnerPrefixIterator` in fd5537a9d736a5f412d93230f5fb50f96a072223

File: benches/benchmarks.rs [hunk 16]
```diff
@@ -14,9 +14,9 @@ pub fn criterion_benchmark(c: &mut Criterion) {
         b.iter(|| {
             let mut store = InMemoryStore::new();
             let mut updater = Updater::new(&mut store, EMPTY_TRIE_ROOT);
-            let mut x = a;
+            let mut x: u128 = a;
             for _ in 0..10_000 {
-                x = (x * a) % m;
+                x = x.saturating_add(a) % m;
```
Comment (afck) [inline benches/benchmarks.rs:19]: Did you mean `saturating_mul`?

Comment (afck) [reply benches/benchmarks.rs:19]: Wait, even that wouldn't work: This _is_ going to saturate `u128` often, so a lot of values will be `u128::MAX % m`.

We'd need 256 bits for this.

File: src/wire_trie.rs [hunk 17]
```diff
@@ -36,29 +36,28 @@ const VERSION: u8 = 0;
 ///
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/wire_trie.rs:36]: So if you have 8 bits and use 1 bit to indicate if something is a leaf or not, then you have 7 bits left.

You don't really need to know the number of 1s in the bitmask. You know there are always 32 bytes in the mask and you can just count the 1s with `.count_ones()` to find a branch's index.

When you have 31 branches or fewer, you need to know the number so you can know where the sorted index of branch indices ends.  You only need 5 bits for that.  You can hand that slice to do a binary search for the branch index when you are reading a key.

But you are revealing that this is inefficient.  If the five length bits are `0b11111`, then we know we have a bitmask rather than a sorted index of branch indices.  This means we only need 6 bits to characterize if a node is a leaf, radix-31 or radix-256 node.

Having 2 bits to spare is convenient.  I would like to have a bit indicating if the node is being marked for keeping with garbage collection. I would like another bit to indicate if all of a nodes descendants are present for fast-syncing.

[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/wire_trie.rs:36]: On second thought, since we want to have an optional value associated with each node, we can't use `0b00000` to indicate a bitmask.  It will be possible to have a node with 0 branches.

We'll probably need another byte for doing bookkeeping in fast-sync/garbage collection.

File: src/wire_trie.rs [hunk 18]
```diff
@@ -73,31 +72,31 @@ impl<'a> WireTrieLeafRef<'a> {
...
+pub(crate) enum TrieLeafOrBranch<'a> {
+    Leaf(Leaf<'a>),
+    Branch(&'a Digest),
+    KeyNotFound,
+    IndexOutOfRange,
 }
 
-pub(crate) struct TrieProof<'a> {
+pub(crate) struct Proof<'a> {
     version_byte_and_envelop_hash: Digest,
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/wire_trie.rs]: Fixed in fa96db639e6561ceb6d771d7e1612e38cc49bd7f

File: src/wire_trie.rs [hunk 19]
```diff
@@ -117,8 +121,8 @@ impl<'a> WireTrieRef<'a> {
     /// - If the trie is not a node this returns 0 bytes.
     fn search_index_length(&self) -> usize {
         match self.tag() {
-            NODE31_TAG => self.0[0] as usize & 0b11111,
-            NODE256_TAG => 32,
+            Tag::Node31 => self.0[0] as usize & 0b11111,
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/wire_trie.rs]: Fixed in 8505230

File: src/wire_trie.rs [hunk 20]
```diff
@@ -230,48 +238,91 @@ impl<'a> WireTrieRef<'a> {
...
+
+        let affix = self.key_or_affix();
+
+        // If the affix is not prefix of the keys bytes remaining, then return KeyNotFound
+        if search_key.len() <= *key_bytes_read as usize + affix.len()
+            || !affix
+                .iter()
+                .zip(&search_key[*key_bytes_read as usize..])
+                .all(|(affix_byte, key_byte)| affix_byte == key_byte)
+        {
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/wire_trie.rs]: Fixed in 89e2848

File: src/wire_trie.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/wire_trie.rs]: Should be fixed in d710b4e1f0c7a85e3a09cb54cc3fea44d26ed713

File: src/store/updater.rs (hunk 1 above)
Comment (xcthulhu) [reply src/store/updater.rs]: I made drastic changes in 8505230 to fix this. In particular, I ended up making an iterator over the branch indices / digest pairs that hid the details of whether the node was radix-31 or radix-256.

This code just uses accessors on `Trie` now.

File: src/store.rs (hunk 2 above)
Comment (xcthulhu) [reply src/store.rs]: I think so?  However we are going to ultimately need to support corrupted tries for resiliency.  There is no reason a node with a corrupted historical trie should crash.  This code should never panic.

File: src/store.rs (hunk 3 above)
Comment (xcthulhu) [reply src/store.rs]: Should be fixed in dc79b08

File: src/store.rs (hunk 4 above)
Comment (xcthulhu) [reply src/store.rs]: Should be fixed in https://github.com/casper-network/casper-trie/commit/dc79b080a66c3b41e37b2a9bfd8c83bcff7e4fad

File: benches/benchmarks.rs (hunk 16 above)
Comment (xcthulhu) [reply benches/benchmarks.rs]: *sigh* I read the data-sheet incorrectly.  A proper value for `a` in a linear congruential random number generator with this modulus is 13891176665706064842. I've fixed this in eb0a3be.

File: src/wire_trie.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (afck) [reply src/wire_trie.rs]: Yes, when I wrote this I hadn't thought of the possibility of having an optional value at every node yet.

File: src/wire_trie.rs [hunk 21]
```diff
@@ -341,4 +342,93 @@ impl<'a> Trie<'a> {
...
+    }
+}
+
+pub(crate) struct TrieBranchIndexIterator<'a> {
+    branch_indices_or_bitvector: &'a [u8],
+    tag: TrieTag,
+    branch_index_length: u8,
+    index: u8,
+    current_bit_vector: Option<u64>,
+}
```
Comment (afck) [inline src/wire_trie.rs:377]: I'd be tempted to try and simplify this :golf::

```rust
pub(crate) enum TrieBranchIndexIterator<'a> {
    Node31(slice::Iter<'a, u8>),
    Node256 {
        bitvector: &'a [u8],
        index: u8,
        current_bit_vector: u64,
    }
}
```

Otherwise it's pretty confusing that not just `branch_indices_or_bitvector` but also `index` has a completely different meaning in both cases.

Comment (afck) [inline src/wire_trie.rs:377]: Would be nice to have a constructor for this, to hide the implementation details.

File: src/wire_trie.rs [hunk 22]
```diff
@@ -1,3 +1,34 @@
...
+//!     - 2: The trie is a radix-256 node.
+//!   - A branch count encoded as the least significant 5 bits when this is a Node31
+//! - Byte 1: Key or affix length.
+//! - Bytes 2 to at most 256+2: The key or affix. These bytes are for a key if the highest three bits
+//!   in byte 1 were LEAF_TAG (ie, 0). Otherwise it is an affix, because the trie is a node with
+//!   branches.
+//! - Additional bytes: Depends on the tag
+//!   - Leaf: these are the values of the leaf, represented as a blob of bytes with no prefixed length.
+//!   - Node31: this type has
+//!     - Search index bytes, with length indicated by the bottom 5 bits of the first byte. These bytes are always in order so we can use binary searches.
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/wire_trie.rs]: Should be fixed in 97fd16ac9f4e4a39f9a26da2cac4ebf5ee822c8a

File: src/wire_trie.rs [hunk 23]
```diff
@@ -1,3 +1,34 @@
...
+//!     - A list of hashes. Each hash is DIGEST_LENGTH (ie, 32) bytes long. The length of this
+//!       list is the same as the bottom 5 bits of the first byte.
+//!   - NODE32_TYPE: this type has
```
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/wire_trie.rs]: Right, fixed this in 97fd16ac9f4e4a39f9a26da2cac4ebf5ee822c8a

File: src/store/backends/in_memory.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (xcthulhu) [reply src/store/backends/in_memory.rs]: Even if it's not used for multiple backend implementations, it's better to have a `new` constructor since now the callsite doesn't need to know if the implementation is an `enum` or `struct`.

Fixed in 97fd16ac9f4e4a39f9a26da2cac4ebf5ee822c8a

File: src/wire_trie.rs (hunk 21 above)
Comment (xcthulhu) [reply src/wire_trie.rs:374]: Should be fixed in ccde49acbc7f65a50bf546539765673b594c7ac7

Comment (xcthulhu) [reply src/wire_trie.rs:390]: I went ahead and made these changes in cca212a8c44035983cd138497a33dc747185e0cd
